File: sdk/typescript/src/resources/memories.ts
/** CreateMemory/GetMemory/UpdateMemory/DeleteMemory/Archive/Restore/List —
 * the Phase 1 canonical operations (05-api-design.md).
 */

import type { Transport } from "../http.js";
import type {
  CreateMemoryRequest,
  MemoryObject,
  MemoryRecord,
  MemoryState,
  Page,
  RelationshipsView,
  UpdateMemoryRequest,
  VersionInfo,
} from "../types.js";

export class MemoriesResource {
  constructor(private readonly transport: Transport) {}

  async create(
    request: CreateMemoryRequest,
    options: { idempotencyKey?: string } = {},
  ): Promise<MemoryObject> {
    const headers = options.idempotencyKey
      ? { "Idempotency-Key": options.idempotencyKey }
      : undefined;
    return this.transport.request<MemoryObject>("POST", "/v1/memories", {
      json: request,
      headers,
    });
  }

  async get(memoryId: string, options: { version?: number } = {}): Promise<MemoryObject> {
    return this.transport.request<MemoryObject>("GET", `/v1/memories/${memoryId}`, {
      params: { version: options.version },
    });
  }

  /** Version history (lightweight `VersionInfo` rows, not full Memory
   * Objects — fetch a specific version via `.get(memoryId, { version })`.
   */
  async listVersions(memoryId: string): Promise<VersionInfo[]> {
    const data = await this.transport.request<{ memory_id: string; versions: VersionInfo[] }>(
      "GET",
      `/v1/memories/${memoryId}/versions`,
    );
    return data.versions;
  }

  /** Filtered read of lifecycle-summary records (not full Memory Objects —
   * this is a projection listing, not Search). Fetch a specific memory_id
   * via `.get()` for the full object.
   */
  async list(
    options: {
      namespace?: string;
      state?: MemoryState;
      limit?: number;
      continuationToken?: string;
    } = {},
  ): Promise<Page<MemoryRecord>> {
    return this.transport.request<Page<MemoryRecord>>("GET", "/v1/memories", {
      params: {
        namespace: options.namespace,
        state: options.state,
        limit: options.limit,
        continuation_token: options.continuationToken,
      },
    });
  }

  async update(
    memoryId: string,
    request: UpdateMemoryRequest,
    options: { expectedVersion: number; idempotencyKey?: string },
  ): Promise<MemoryObject> {
    const headers: Record<string, string> = { "If-Match": String(options.expectedVersion) };
    if (options.idempotencyKey) headers["Idempotency-Key"] = options.idempotencyKey;
    return this.transport.request<MemoryObject>("PATCH", `/v1/memories/${memoryId}`, {
      json: request,
      headers,
    });
  }

  async delete(memoryId: string): Promise<Record<string, unknown>> {
    return this.transport.request("DELETE", `/v1/memories/${memoryId}`);
  }

  async archive(memoryId: string): Promise<MemoryObject> {
    return this.transport.request<MemoryObject>("POST", `/v1/memories/${memoryId}/archive`);
  }

  async restore(memoryId: string): Promise<MemoryObject> {
    return this.transport.request<MemoryObject>("POST", `/v1/memories/${memoryId}/restore`);
  }

  /** Graph edges touching this memory, outbound and inbound (Phase 4 task 1,
   * ADR-0006) — a read-only view over the relationship-graph projection.
   */
  async relationships(memoryId: string): Promise<RelationshipsView> {
    return this.transport.request<RelationshipsView>(
      "GET",
      `/v1/memories/${memoryId}/relationships`,
    );
  }
}

File: sdk/python/mip_sdk/resources/memories.py
"""CreateMemory/GetMemory/UpdateMemory/DeleteMemory/Archive/Restore/List —
the Phase 1 canonical operations (05-api-design.md).
"""

from __future__ import annotations

from typing import Any

from mip_sdk._http import Transport
from mip_sdk.models.memory import MemoryObject, MemoryRecord, MemoryState, VersionInfo
from mip_sdk.models.requests import CreateMemoryRequest, UpdateMemoryRequest
from mip_sdk.models.retrieval import Page


class MemoriesResource:
    def __init__(self, transport: Transport) -> None:
        self._transport = transport

    def create(
        self, request: CreateMemoryRequest, *, idempotency_key: str | None = None
    ) -> MemoryObject:
        headers = {"Idempotency-Key": idempotency_key} if idempotency_key else None
        data = self._transport.request(
            "POST",
            "/v1/memories",
            json=request.model_dump(mode="json", exclude_none=True),
            headers=headers,
        )
        return MemoryObject.model_validate(data)

    def get(self, memory_id: str, *, version: int | None = None) -> MemoryObject:
        params = {"version": version} if version is not None else None
        data = self._transport.request("GET", f"/v1/memories/{memory_id}", params=params)
        return MemoryObject.model_validate(data)

    def list_versions(self, memory_id: str) -> tuple[VersionInfo, ...]:
        """Version history (lightweight `VersionInfo` rows, not full Memory
        Objects — fetch a specific version via `.get(memory_id, version=N)`.
        """
        data = self._transport.request("GET", f"/v1/memories/{memory_id}/versions")
        return tuple(VersionInfo.model_validate(v) for v in data["versions"])

    def list(
        self,
        *,
        namespace: str | None = None,
        state: MemoryState | None = None,
        limit: int | None = None,
        continuation_token: str | None = None,
    ) -> Page[MemoryRecord]:
        """Filtered read of lifecycle-summary records (not full Memory
        Objects — this is a projection listing, not Search; storage/interfaces.py
        `MemoryRecord`). Fetch a specific memory_id via `.get()` for the full object.
        """
        params: dict[str, Any] = {
            "namespace": namespace,
            "state": state.value if state else None,
            "limit": limit,
            "continuation_token": continuation_token,
        }
        data = self._transport.request(
            "GET", "/v1/memories", params={k: v for k, v in params.items() if v is not None}
        )
        return Page[MemoryRecord](
            items=tuple(MemoryRecord.model_validate(item) for item in data["items"]),
            complete=data["complete"],
            continuation_token=data["continuation_token"],
        )

    def update(
        self,
        memory_id: str,
        request: UpdateMemoryRequest,
        *,
        expected_version: int,
        idempotency_key: str | None = None,
    ) -> MemoryObject:
        headers = {"If-Match": str(expected_version)}
        if idempotency_key:
            headers["Idempotency-Key"] = idempotency_key
        data = self._transport.request(
            "PATCH",
            f"/v1/memories/{memory_id}",
            json=request.model_dump(mode="json", exclude_none=True),
            headers=headers,
        )
        return MemoryObject.model_validate(data)

    def delete(self, memory_id: str) -> dict[str, Any]:
        result: dict[str, Any] = self._transport.request("DELETE", f"/v1/memories/{memory_id}")
        return result

    def archive(self, memory_id: str) -> MemoryObject:
        data = self._transport.request("POST", f"/v1/memories/{memory_id}/archive")
        return MemoryObject.model_validate(data)

    def restore(self, memory_id: str) -> MemoryObject:
        data = self._transport.request("POST", f"/v1/memories/{memory_id}/restore")
        return MemoryObject.model_validate(data)
